torch/_inductor/codegen/wrapper.py

```diff
@@ -1958,28 +1972,96 @@ def codegen_unbacked_symbol_decl(self, symbol):
+        #
+        # However this does not work with cpp wrapper. With cpp wrapper, we make
+        # two passes and the kernels are shared from the first pass to the next.
+        # Therefore, both the Python and CppWrapper need to share the some
+        # codegen infra for cpp_wrapper=True case. Therefore for cpp_wrapper, we
+        # fallback to the old way of inlining subgraphs in the output code. Once
+        # we update CppWrapperCpu, we can remove this function.
+        def _codegen_subgraph_prefix():
+            for inner_input, outer_input in zip(
+                subgraph.graph.graph_inputs, outer_inputs
+            ):
+                self.writeline(
+                    f"{self.declare}{inner_input} = {outer_input}{self.ending}"
+                )
+
+        def _codegen_subgraph_suffix():
+            for inner_output, outer_output in zip(
+                subgraph.graph.graph_outputs, outer_outputs
+            ):
+                self.writeline(
+                    f"{outer_output} = {inner_output.codegen_reference()}{self.ending}"
+                )
+
+        try:
+            self.push_codegened_graph(subgraph.graph)
+            self.writeline(f"{self.comment} subgraph: {subgraph.name}")
+            _codegen_subgraph_prefix()
+            parent_graph = V.graph
+            with V.set_graph_handler(subgraph.graph):
+                subgraph.graph.codegen_subgraph(
+                    parent_graph=parent_graph,
+                )
+            _codegen_subgraph_suffix()
+        finally:
+            self.pop_codegened_graph()
+
     def codegen_subgraph_prefix(self, subgraph, outer_inputs, outer_outputs):
+        subgraph.graph.add_symbol_graph_inputs()
         for inner_input, outer_input in zip(subgraph.graph.graph_inputs, outer_inputs):
             self.writeline(f"{self.declare}{inner_input} = {outer_input}{self.ending}")
 
     def codegen_subgraph_suffix(self, subgraph, outer_inputs, outer_outputs):
         for inner_output, outer_output in zip(
-            subgraph.graph.graph_outputs, outer_outputs
+            subgraph.graph.get_output_names(), outer_outputs
         ):
-            self.writeline(
-                f"{outer_output} = {inner_output.codegen_reference()}{self.ending}"
-            )
+            self.writeline(f"{outer_output} = {inner_output}{self.ending}")
+
+    def codegen_subgraph_call(self, subgraph, outer_inputs, outer_outputs):
+        # Get the input and output names of the subgraph
+        inner_inputs = ", ".join(subgraph.graph.graph_input_names)
+        if len(subgraph.graph.graph_input_names) == 1:
+            inner_inputs += ","
+
+        output_names = subgraph.graph.get_output_names()
+        inner_outputs = ", ".join(output_names)
+        if len(output_names) == 1:
+            inner_outputs += ","
+
+        # Call the subgraph launcher function
+        self.writeline(f"({inner_outputs}) = {subgraph.graph.name}([{inner_inputs}])")
```

Comment:
somewhat arbitrarily commenting here:

https://www.internalfb.com/phabricator/paste/view/P1632946348?lines=166%2C166%2C167%2C171%2C172

We are deleting `arg1_1` and other inputs to the subgraphs too late. We should create the list that would be the input to the subgraph, then delete `arg1_1`. 

Similarly, the arguments to the subgraphs should be deleted in the invoking code. 

```
            # subgraph: false_graph_0
            false_graph_0_arg0_1 = reinterpret_tensor(arg1_1, ((-1) + s0, 20), (20, 1), 0)
            false_graph_0_arg1_1 = reinterpret_tensor(arg2_1, ((-1) + s0, 20), (20, 1), 0)
            (false_graph_0_buf0, false_graph_0_buf1) = false_graph_0([false_graph_0_arg0_1, false_graph_0_arg1_1, s0])
            buf0[0] = false_graph_0_buf0
            buf0[1] = false_graph_0_buf1
        del arg1_1
        del arg2_1
````

-> 
```
false_graph_0_arg0_1 = reinterpret_tensor(arg1_1, ((-1) + s0, 20), (20, 1), 0)
false_graph_0_arg1_1 = reinterpret_tensor(arg2_1, ((-1) + s0, 20), (20, 1), 0)
false_graph_0_args = [[false_graph_0_arg0_1, false_graph_0_arg1_1, s0]]
del arg1_1
del arg2_1
del false_graph_0_arg0_1
del false_graph_0_arg1_1
(false_graph_0_buf0, false_graph_0_buf1) = false_graph_0(false_graph_0_args)
```

Otherwise we are keeping those tensors alive longer than they need to be.

Reply:
I agree with `false_graph_0_arg1_1` and `false_graph_0_arg1_1` to be deleted right away after creating the `list`. I will send a fix for that.

But `arg1_1` and `arg2_1` might be needed later on. We will have to liveness analysis of these variables, and only for cases when there are no more uses for these args, we can safely delete them. Although the optimization makes sense, I am little bit hesitant to do it in this PR (the current implementation on `main` also has same unoptimized version).

Reply:
As it is a pre-existing issue, we can file issue or fix in follow up.